homeassistant/components/reolink/host.py

```diff
@@ -363,22 +363,22 @@ async def handle_webhook(
         self, hass: HomeAssistant, webhook_id: str, request: Request
     ):
         """Shield the incoming webhook callback from cancellation."""
-        await asyncio.shield(self.handle_webhook_shielded(hass, webhook_id, request))
```

Comment:
You probably want to keep doing the work in the shielded task since you could get cancelation when awaiting request.text (unless you want it there)

Reply:
@bdraco does that mean I can catch the asyncio.CancelledError reliably? or does that not work since the cancellation comes from higher up?

I could potentially catch the `asyncio.CancelledError` from the `request.text()` and issue to poll instead (No other choise if the data is lost).

However I do want to have the `request.text()` before the `asyncio.shield`, In testing I am losing about 50% of pushes when I put the `request.text()` inside the shield, while I did not observe any losses when having it before the `asyncio.shield`.
This is either a race condition because the `asyncio.shield` can just take 100 ms, or there is something more complex in aiohttp. In the [aiohttp documentation](https://docs.aiohttp.org/en/stable/web_advanced.html) I read:

> Reading from the WebSocket (await ws.receive()) must only be done inside the request handler task;

Which made me think there might be some more complex reason why I schould not put the `request.text()` in a diffrent task (`asyncio.shield`).

But you @bdraco are a legend in this kind of stuff, so please teach me how this works.

Reply:
Here is the shield implementation https://github.com/python/cpython/blob/4fa5fda14b11457dda7ef389e5486bfe3ea7b8f5/Lib/asyncio/tasks.py#L841

As long as the shield call runs anything its shielding should get run.

Reply:
I have no way to test this but, I think you can simplify the code a bit https://github.com/bdraco/home-assistant/commit/5a19f2e786ab93d19b8eef049ad380ea57db67e9

Reply:
Yeah I agree with bdraco. The shielding is not the thing that is adding the delay as per the source, it's merely adding an extra future, which means there is a small bit of extra execution logic at the end.

aiohttp sadly doesn't have an option where we can pro-actively fetch the content of an incoming request, however an iteration of the event loop shouldn't take that long that Reolink hangs up, unless they have a very tiny timeout?

Reply:
@balloob from testing (see image in PR desciption) the timeout of the camera is shorter than 68 ms, I think that is a very tiny timeout, but I am not so deep into aiohttp to know.

So I think what is going on is that the `await asyncio.shield(self.handle_webhook_shielded(hass, webhook_id, request))` yields to the event loop before executing the `handle_webhook_shielded`, and if the event loop is doing a lot of stuff the iteration of the event loop can apperently take 68 or even 182 ms? Please do correct me if I am wrong, I am very out of my comfort zone here.

@bdraco proposes to do:
```
shielded_future = asyncio.shield(self._handle_webhook(hass, webhook_id, request))
await shielded_future
```
Does this change the behaviour since the await is now seperated?
Does this mean self._handle_webhook will get excuted withouth yielding to the event loop first?

My appologies for beeing ignorant, I still need to learn a lot about aiohttp/asyncio

Reply:
@bdraco thank you for your example, I have largely followed that.
I have tested extensevely and it now seems to work good.
Most of the time there is only a 1 ms delay between the `handle_webhook` and shielded `_handle_webhook`.

It turns out it is really hard to predict when the Reolink camera will drop its connection.
When multiple pushes come in very close to each other (within 100 ms) because for instance the cam sees motion and just after that recognizes its a person (motion and motion+person event very close together) It will drop its connection really fast (also on the second call.

But anyway, the current code has the polling backup which in testing works flawlessly.

Reply:
After this additonal testing I agree with @balloob and @bdraco that it is better to first shield and then in the shielded task attempt to read the data.